autoloader.php

```diff
@@ -0,0 +1,153 @@
+<?php
+
+/**
+ * Autoload the classes used by the AMP plugin.
+ *
+ * Class AMP_Autoloader
+ */
+class AMP_Autoloader {
+
+	/**
+	 * Map of Classname to relative filepath sans extension.
+	 *
+	 * @example Format:
+	 *
+	 *  array(
+	 *      'Class_Name1' =>  'subdir-of-includes/filename1',
+	 *      'Class_Name2' =>  '2nd-subdir-of-includes/filename2',
+	 *      'Class_Name3' =>  '!/full/path/to/filename3.php',
+	 *  );
+	 *
+	 * @var string[]
+	 */
+	private $_classmap;
+
+	/**
+	 * Provide access to the singlton instance
+	 *
+	 * @var AMP_Autoloader
+	 */
+	private static $_instance;
+
+	/**
+	 * Perform the autoload.
+	 *
+	 * Run as few lines of code as possible.
+	 *
+	 * If $class_name has first char as '!' strip it and assume a full path w/extension.
+	 * Otherwise assume subdir or /includes/ w/o extension.
+	 *
+	 * @param string $class_name
+	 */
+	function _autoload( $class_name ) {
+		if ( ! isset( $this->_classmap[ $class_name ] ) ) {
+			return;
+		}
+		$filepath = $this->_classmap[ $class_name ];
+		require( '!' !== $filepath[ 0 ] ? __DIR__ . "/includes/{$filepath}.php" : substr( $filepath, 1 ) );
+	}
+
+	/**
+	 * Trigger creation of AMP_Autoloader singleton instance and return it.
+	 *
+	 * @return AMP_Autoloader
+	 */
+	public static function instance() {
```

Comment:
There aren't any singletons in this plugin, so I think static methods should be used instead for consistency.

Reply:
Funny, I debated using static but assumed you'd prefer instances. :-)

Reply:
Normally I'd prefer non-singleton instances, but given most classes are just being used as namespaces with static methods, using static here again seems to be most in keeping with everything else.

Reply:
Updated to use static methods.